cores/arduino/stm32/analog.c

```diff
@@ -434,25 +444,114 @@ void HAL_ADC_MspInit(ADC_HandleTypeDef *hadc)
+  }
+ #endif
+  
+#ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[2]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[2]);
+  }
+#endif
+  
+#ifdef ADC4_BASE
+  if (hadc->Instance == ADC4)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[3]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[3]);
+  }
+#endif
+
   /* ADC Channel GPIO pin configuration */
-  GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pin);
 #ifdef GPIO_MODE_ANALOG_ADC_CONTROL
   GPIO_InitStruct.Mode = GPIO_MODE_ANALOG_ADC_CONTROL;
 #else
   GPIO_InitStruct.Mode = GPIO_MODE_ANALOG;
 #endif
   GPIO_InitStruct.Pull = GPIO_NOPULL;
   HAL_GPIO_Init(port, &GPIO_InitStruct);
+  
+  if (use_dma)
+  {
+    /*##-3- Ask DMA informations ###############################################*/
+    dma_allocate(hadc, ADC_PERIPH, &channel, &irqHandler, &indice);
+
+    /*##-4- Configure the DMA ##################################################*/
+        
+    #if defined(STM32L0xx) || defined(STM32L4xx)
+    DmaHandleTab[indice].Init.Request             = DMA_REQUEST_0;
+    #endif
+
+    #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+    #ifdef ADC1_BASE
+    if (hadc->Instance == ADC1)
+    {
+      DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_0;
+    }
+    #endif
+    #ifdef ADC2_BASE
+    if (hadc->Instance == ADC2)
+    {
+      DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_1;
+    }
+    #endif
+    #ifdef ADC3_BASE
+    if (hadc->Instance == ADC3)
+    {
+      DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_2;
+    }
+    #endif
+    #endif //STM32F2xx STM32F4xx STM32F7xx
+
+    DmaHandleTab[indice].Instance                 = channel;
+    DmaHandleTab[indice].Init.Direction           = DMA_PERIPH_TO_MEMORY;
+    DmaHandleTab[indice].Init.PeriphInc           = DMA_PINC_DISABLE;
+    DmaHandleTab[indice].Init.MemInc              = DMA_MINC_ENABLE;
+    DmaHandleTab[indice].Init.PeriphDataAlignment = DMA_PDATAALIGN_HALFWORD;
+    DmaHandleTab[indice].Init.MemDataAlignment    = DMA_MDATAALIGN_WORD;
+    DmaHandleTab[indice].Init.Mode                = DMA_NORMAL;
+    DmaHandleTab[indice].Init.Priority            = DMA_PRIORITY_MEDIUM;
+
+    #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+    DmaHandleTab[indice].Init.FIFOMode              = DMA_FIFOMODE_DISABLE;
+    DmaHandleTab[indice].Init.FIFOThreshold         = DMA_FIFO_THRESHOLD_HALFFULL;
+    DmaHandleTab[indice].Init.MemBurst              = DMA_MBURST_SINGLE;
+    DmaHandleTab[indice].Init.PeriphBurst           = DMA_PBURST_SINGLE;
+    #endif //STM32F2xx STM32F4xx STM32F7xx
+        
+    /* Deinitialize  & Initialize the DMA for new transfer */
+    HAL_DMA_DeInit(&DmaHandleTab[indice]);
+    HAL_DMA_Init(&DmaHandleTab[indice]);
+
+    /* Associate the DMA handle */
+    __HAL_LINKDMA(hadc, DMA_Handle, DmaHandleTab[indice]);
```

Comment:
where is DMA_Handle defined ?

Reply:
DMA_Handle is part of the struct ADC_HandleTypeDef. You can see it in stm32f3xx_hal_adc.h for example.

Reply:
ok - I got confused by the way __HAL_LINKDMA macro is built. thanks